src/ynEIGEN/RedemptionAssetsVault.sol

```diff
@@ -0,0 +1,252 @@
+     * @dev Requires the caller to be the redeemer and the contract to not be paused.
+     * @dev Iterates over the supported assets, transferring each asset's balance to the user until fully depleted.
+     */
+    function transferRedemptionAssets(address to, uint256 amount, bytes calldata /* data */) public onlyRedeemer whenNotPaused nonReentrant {
+        uint256 balance = availableRedemptionAssets();
+        if (balance < amount) revert InsufficientAssetBalance(ETH_ASSET, amount, balance);
+
+        uint256 amountTransferred = amount;
+        IERC20[] memory assets = assetRegistry.getAssets();
+        for (uint256 i = 0; i < assets.length; ++i) {
+            IERC20 asset = assets[i];
+            uint256 assetBalance = balances[address(asset)];
+            if (assetBalance > 0) {
+                uint256 assetBalanceInUnit = assetRegistry.convertToUnitOfAccount(asset, assetBalance);
+                if (assetBalanceInUnit >= amount) {
+                    uint256 reqAmountInAsset = assetRegistry.convertFromUnitOfAccount(asset, amount);
+                    IERC20(asset).safeTransfer(to, reqAmountInAsset);
+                    balances[address(asset)] -= reqAmountInAsset;
+                    break;
+                } else {
+                    IERC20(asset).safeTransfer(to, assetBalance);
+                    balances[address(asset)] = 0;
+                    amount -= assetBalanceInUnit;
+                }
+            }
+        }
+
+        emit AssetTransferred(ETH_ASSET, msg.sender, to, amountTransferred);
+    }
+
+    /** 
+     * @notice Withdraws a specified amount of redemption assets and processes them through ynETH.
+     * @param amount The amount of ETH to withdraw and process.
+     * @dev Requires the caller to be the redeemer and the contract to not be paused.
+     */
+    function withdrawRedemptionAssets(uint256 amount) public onlyRedeemer whenNotPaused nonReentrant {
+        IERC20[] memory assets = assetRegistry.getAssets();
+        for (uint256 i = 0; i < assets.length; ++i) {
+            IERC20 asset = assets[i];
+            uint256 assetBalance = balances[address(asset)];
+            if (assetBalance > 0) {
+                uint256 assetBalanceInUnit = assetRegistry.convertToUnitOfAccount(asset, assetBalance);
+                if (assetBalanceInUnit >= amount) {
+                    uint256 reqAmountInAsset = assetRegistry.convertFromUnitOfAccount(asset, amount);
+                    ynEigen.processWithdrawn(reqAmountInAsset, address(asset));
+                    balances[address(asset)] -= reqAmountInAsset;
+                    break;
+                } else {
+                    ynEigen.processWithdrawn(assetBalance, address(asset));
+                    balances[address(asset)] = 0;
+                    amount -= assetBalanceInUnit;
+                }
+            }
+        }
+        emit AssetWithdrawn(ETH_ASSET, msg.sender, address(ynEigen), amount);
```

Comment:
let's apply the same rules here as in the other function in terms of even emission?

Reply:
https://github.com/yieldnest/yieldnest-protocol/pull/164/commits/ba302f104a474d7d6a32dd61bbe2168b9c12c56a